src/protocol_edge.c

```diff
@@ -102,6 +152,15 @@ bool add_edge_h(meshlink_handle_t *mesh, connection_t *c, const char *request) {
 		from = new_node();
 		from->status.blacklisted = mesh->default_blacklist;
 		from->name = xstrdup(from_name);
+
+		from->submesh = NULL;
+
+		if(0 != strcmp(from_submesh_name, "0")) {
```

Comment:
Since you are repeating this pattern a number of times, it might make sense to create a function called `lookup_or_create_submesh(mesh, const char *submesh_name)`, that does all these things in one go.

Reply:
I created lookup_or_create_submesh(mesh, const char *submesh_name) as you suggested and replaced all the similar steps with this function